Primeiramente devemos importar os nossos dados coletados, armazenando ele no dataframe Df que será tratado

In [50]:
import pandas as pd

# Carregar o arquivo CSV
file_path = 'C:\\Users\\delve\\OneDrive\\Eu\\GitHub\\PO-233-Aprendizado-de-Maquina\\Coleta de dados\\saida.csv'
df = pd.read_csv(file_path, index_col=0)

# Exibir as primeiras linhas do dataframe para entender sua estrutura
# df.head()

Etapas que iremos realizar no pré-processamento:
* Transformação de todos nossos dados em floats
* Eliminação de Colunas desnecessárias, que não agregam para o problema
* Eliminação de Colunas que possuem mais que 30% dos seus dados vazios
* Normalização dos dados, assim como normalização da nossa performance

# Eliminação de colunas desnecessárias

Vamos então primeiro mostrar as colunas que possuem valores ausentes, para realizar o tratamento.


Além disso, vamos tornar todos os valores da nossa tabela em numéricos. Caso não sejam numéricos, estarão classificados como NaN

In [51]:
for col in df.columns:
    if df[col].dtype == 'object':
        df[col] = pd.to_numeric(df[col], errors='coerce')

# Mostrar informações sobre valores ausentes
missing_data = df.isnull().sum()

# Exibir as primeiras linhas após conversão
# print(missing_data)
len(df.columns)

77

In [52]:
# Definir um limite de 10% de valores ausentes
# limite_ausentes = 0.1 * len(df)
missing_percentage = df.isnull().mean()

# Remover colunas com mais de 10% de valores ausentes
# df = df.dropna(thresh=limite_ausentes, axis=1)
cols_to_drop = missing_percentage[missing_percentage > 0.1].index

# Preencher os valores ausentes restantes com a mediana da coluna
# 
df_cleaned = df.drop(columns=cols_to_drop)


# Verificar o resultado
# df
df = df_cleaned
df = df.fillna(df.median())
df.isnull().sum().sum()  # Verificando se ainda há valores ausentes
df = df.drop(df.columns[0], axis = 1)
df.to_csv('Non_Normalized_Data.csv')
df

Sales & Services Revenue   Gross Profit   Selling, General & Admin    \
                                                                             
0                         78.52          95.46                       32.97   
1                         65.96          70.60                       37.74   
2                         44.58          56.70                       32.12   
3                          9.20          -6.35                        7.87   
4                          6.95           9.69                       10.74   
...                         ...            ...                         ...   
149                       14.03          15.17                        3.94   
150                       13.65          16.89                        6.97   
151                       17.53          19.52                        2.08   
152                       17.96          17.06                        9.92   
153                        6.88           7.69                        9.41   

      Non-Operating (Income) Loss   Pretax Income   \
                                                     
0                           -52.45          132.21   
1                           167.74           84.49   
2                            25.78           63.03   
3                           121.46          -10.06   
4                           -15.22            6.64   
...                            ...             ...   
149                         -48.52           19.78   
150                         -89.44           21.40   
151                        1440.26           34.06   
152                         -71.92           17.74   
153                         136.64            6.68   

      Income (Loss) from Cont Ops   Basic Weighted Avg Shares   \
                                                                 
0                           145.67                        1.84   
1                            84.99                        1.63   
2                            60.99                        1.14   
3                           -11.25                       -1.01   
4                             6.68                       -6.05   
...                            ...                         ...   
149                         136.80                       -0.35   
150                          12.85                       -0.82   
151                          38.37                       -0.83   
152                          18.72                       -0.68   
153                          -0.52                       -0.67   

      Basic EPS, GAAP    Diluted Weighted Avg Shares   Diluted EPS, GAAP    \
                                                                             
0                141.23                          1.95               140.97   
1                 82.02                          1.29                82.63   
2                 59.18                          0.93                59.51   
3                -10.34                         -1.45                -9.95   
4                 13.54                         -6.12                13.63   
...                 ...                           ...                  ...   
149              137.63                         -0.53               138.05   
150               13.78                         -0.90                13.87   
151               39.52                         -0.98                39.73   
152               19.52                         -0.89                19.79   
153                0.15                         -0.90                 0.39   

      ...  Net Change in LT Investment   Cash from Investing Activities   \
      ...                                                                  
0     ...                        -31.30                           -20.53   
1     ...                        192.65                           191.75   
2     ...                         18.37                            19.32   
3     ...                        -

Número de colunas antes do tratamento: 77

Número de colunas após o tratamento: 57

Substituimos os valores faltantes restantes pela média dos valores 

# Eliminação de Outliers

* Antes de normalizar meus valores, vamos remover os outliers, para isso, primeiro vamos calcular o z-score de cada elemento, isso é, o valor de cada elemento caso a coluna fosse normalizada entre -1 e 1. Com o z-score calculado, vamos escolher um limite para o qual escolheremos os valores nessa nosva distruibuição entre -1 e 1


In [60]:
import numpy as np
data = df
# Calcular z-scores
data_zscores = (data - data.mean()) / data.std()

# Definir limiar de z-score
zscore_threshold = 5

# Filtrar dados
clean_data_zscore = data[(np.abs(data_zscores) < zscore_threshold).all(axis=1)]

# Exibir as primeiras linhas dos dados limpos
data = clean_data_zscore
len(data)

117

* Antes da eliminação, estavamos com 157 linhas, após ficamos com apenas 117

# Normalização
Agora o próximo passo é normalizar minhas colunas

valor = valor / std

Escolhi essa forma de normalizar para mantes valores positivos e valores negativos sem perder a distribuição.

In [63]:
from sklearn.preprocessing import MinMaxScaler
"""
# Instanciar o MinMaxScaler

scaler = MinMaxScaler()
scaler_performance = MinMaxScaler(feature_range=(-1, 1))

# Aplicar a normalização
df_normalized = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)


# Separar a coluna Performance
performance_col = df['Performance']
df_normalized = df.drop(columns=['Performance'])

# Normalizar as colunas de 0 a 1
df_normalized = pd.DataFrame(scaler.fit_transform(df_normalized), columns=df_normalized.columns)

# Normalizar a coluna Performance de -1 a 1

performance_normalized = scaler_performance.fit_transform(performance_col.values.reshape(-1, 1))

# Adicionar a coluna Performance normalizada de volta ao dataframe
df_normalized['Performance'] = performance_normalized

# Standardize the data (Z-score normalization)
# Scale the data to fit within the range [-1, 1]
normalized_data = standardized_data / standardized_data.abs().max()
"""
means = data.mean()
stds = data.std()
normalized_data = ((data) / stds)

# Exibir as primeiras linhas dos dados normalizados
data = normalized_data
data.head()

Sales & Services Revenue   Gross Profit   Selling, General & Admin    \
                                                                             
1                      5.266687       4.501444                    1.403538   
2                      3.559565       3.615182                    1.194532   
3                      0.734590      -0.404875                    0.292683   
4                      0.554934       0.617833                    0.399417   
5                      2.224529       2.086859                    0.724455   

      Non-Operating (Income) Loss   Pretax Income   \
                                                     
1                         0.415811        1.407983   
2                         0.063906        1.050363   
3                         0.301087       -0.167645   
4                        -0.037729        0.110652   
5                         0.077143        0.593089   

      Income (Loss) from Cont Ops   Basic Weighted Avg Shares   \
                                                                 
1                         1.104664                    0.545089   
2                         0.792722                    0.381228   
3                        -0.146223                   -0.337754   
4                         0.086824                   -2.023183   
5                         0.456735                   -1.825881   

      Basic EPS, GAAP    Diluted Weighted Avg Shares   Diluted EPS, GAAP    \
                                                                             
1              1.169302                      0.453431             1.178181   
2              0.843688                      0.326892             0.848524   
3             -0.147410                     -0.509671            -0.141872   
4              0.193030                     -2.151162             0.194344   
5              0.612166                     -1.891054             0.610692   

      ...  Net Change in LT Investment   Cash from Investing Activities   \
      ...                                                                  
1     ...                      0.378037                         0.376968   
2     ...                      0.036047                         0.037982   
3     ...                     -0.073449                        -0.058919   
4     ...                     -0.122624                        -0.065171   
5     ...                      0.770378                         0.293376   

      Cash (Repurchase) of Equity   Cash from Financing Activities   \
                                                                      
1                        -0.024700                         0.008630   
2                        -0.055575                        -0.126194   
3                        -9.618249                         0.501440   
4                         0.273286                         0.074960   
5                        -0.060053                        -0.030634   

      Net Changes in Cash   Capital Expenditures   Free Cash Flow   \
                                                                     
1                -0.067022               7.656732         1.025894   
2                -0.088769               0.797234         0.470020   
3                 0.149770              -0.105709         0.094054   
4                -0.060380               0.247365         0.147976   
5                -1.000800               0.520014         0.494994   

      Price/Free Cash Flow   Cash Flow to Net Income   Performance  
                                                                    
1                 -0.189894                  0.053047     1.030579  
2                  0.162383                 -0.092279     0.772541  
3                 -0.169729                  0.110233    -0.868117  
4                  0.128538                  0.025066     1.178573  
5                 -0.113947                  0.004139    -0.092211  

[5 rows x 57 columns]

In [62]:
data.to_csv('Normalized_Data.csv')
data